aquadoggo/src/libp2p/service.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::convert::TryInto;
+
+use anyhow::Result;
+use futures::StreamExt;
+use libp2p::core::muxing::StreamMuxerBox;
+use libp2p::ping::Event;
+use libp2p::swarm::behaviour::toggle::Toggle;
+use libp2p::swarm::{NetworkBehaviour, SwarmBuilder, SwarmEvent};
+use libp2p::{mdns, ping, quic, Multiaddr, PeerId, Transport};
+use log::{info, warn};
+
+use crate::bus::ServiceSender;
+use crate::context::Context;
+use crate::libp2p::Libp2pConfiguration;
+use crate::manager::{ServiceReadySender, Shutdown};
+
+/// Network behaviour for the aquadoggo node.
+#[derive(NetworkBehaviour)]
+struct Behaviour {
+    /// Automatically discover peers on the local network.
+    mdns: Toggle<mdns::tokio::Behaviour>,
+    /// Respond to inbound pings and periodically send outbound ping on every established
+    /// connection.
+    ping: Toggle<ping::Behaviour>,
+}
+
+impl Behaviour {
+    /// Generate a new instance of the composed network behaviour according to
+    /// the libp2p configuration context.
+    fn new(libp2p_config: &Libp2pConfiguration, peer_id: PeerId) -> Result<Self> {
+        // Create an mDNS behaviour with default configuration if the mDNS flag is set
+        let mdns = if libp2p_config.mdns {
+            Some(mdns::Behaviour::new(Default::default(), peer_id)?)
+        } else {
+            None
+        };
+
+        // Create a ping behaviour with default configuration if the ping flag is set
+        let ping = if libp2p_config.ping {
+            Some(ping::Behaviour::default())
+        } else {
+            None
+        };
+
+        Ok(Self {
+            mdns: mdns.into(), // Convert the `Option` into a `Toggle`
+            ping: ping.into(),
+        })
+    }
+}
+
+/// Libp2p service that configures and deploys a network swarm over QUIC transports.
+///
+/// The swarm listens for incoming connections, dials remote nodes, manages
+/// connections and executes predefined network behaviours.
+pub async fn libp2p_service(
+    context: Context,
+    shutdown: Shutdown,
+    tx: ServiceSender,
+    tx_ready: ServiceReadySender,
+) -> Result<()> {
+    // Subscribe to communication bus
+    let mut _rx = tx.subscribe();
+
+    // Read the libp2p configuration parameters from the application context
+    let libp2p_config = context.config.libp2p.clone();
+
+    // Load the libp2p keypair and peer ID
+    let keypair = Libp2pConfiguration::load_or_generate_keypair(context.config.base_path.clone())?;
+    let peer_id = PeerId::from(keypair.public());
+    info!("libp2p peer ID: {peer_id:?}");
```

Comment:
I noticed that this prints as `libp2p peer ID: PeerId("12D3KooWQFPNhLX9h9xjwTEv1xvzkxzTh7PrTw5F82am3r5Xp5gq")` in the logs w. `Debug` (aka w. `PeerId()` type around value) representation. Is it possible to use the `Display` representation instead?